src/nvim/lua/treesitter.c

```diff
@@ -1411,26 +1417,44 @@ static int query_next_capture(lua_State *L)
     push_node(L, capture.node, lua_upvalueindex(2));  // [index, node]
 
     // Now check if we need to run the predicates
+    // a predicate is executed only once within the same match.
     uint32_t n_pred;
     ts_query_predicates_for_pattern(query, match.pattern_index, &n_pred);
 
-    if (n_pred > 0 && (ud->max_match_id < (int)match.id)) {
+    if (n_pred == 0) {
+      // no predicates. Return without the match table as a small perf optimization,
+      // because in the iter_captures() method, match table won't be needed.
+      // [index, node]
```

Comment:
Is this true? The `match` table is still needed for directives, even when there are no predicates.

Besides, we will want the `match` table for every iteration anyway for https://github.com/neovim/neovim/pull/27132.

Reply:
From treesitter's pov, directives and predicates are both predicates. Here is the example of upstream calling `#set!` a predicate. https://tree-sitter.github.io/tree-sitter/syntax-highlighting#unit-testing.

Reply:
Ah, thanks for that clarification. If that is indeed the case then I agree we only need to return the match table when `npred > 0`.

Reply:
Right. When I wrote this earlier, there was no such need other than an *ephemeral* use for predicates/directives. But if #27132 needs `match`, we can remove this *micro-* optimization. We'd need to do profiling and check how significant this would be, because a majority of patterns will have no predicates. See @bfredl's comment: https://github.com/neovim/neovim/issues/27239#issuecomment-1914247646, suggesting this should be done in an `opt-in` manner.